### ISSUE TITLE
[FEATURE] Adaptive length of different parts of blockId 

### ISSUE THREAD

**Issue by zuston** [apache/incubator-uniffle#731]
### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### Search before asking

- [X] I have searched in the [issues](https://github.com/apache/incubator-uniffle/issues?q=is%3Aissue) and found no similar issues.

### Describe the feature

I found some problems about BlockID in production env.
1. Due to too much iterations, some apps' (like XGBoost training using Spark) taskAttempt ID is large, which means the fixed length of 21 is not enough.
2. Due to above problem, I increase the TASK_ATTEMPT_ID_MAX_LENGTH to 23 and decrease the MAX_SEQUENCE_NO to 6. Pity, I found some apps will fail due to the exceeding the  MAX_SEQUENCE_NO range.

### Under the hood of `BLOCKID` construction

BlockID will be created by 3 parts, including partitionId, taskAttemptId and sequenceNumber. If using the fixed length for every part, it's hard to be adapt with different applications.

So why not using the adaptive length of different parts of blockId? As we know, the partition number has been determinated before writing. That means we could adjust the rest 2 parts' length according to the partition number dynamically. 

BTW, another solution is to decrease the MAX partitionID length, the default value of 24 is too large.

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

**Comment by jerqi** [comment]
If we  increase `TASK_ATTEMPT_ID_MAX_LENGTH`, it will cause more GC.

**Comment by zuston** [comment]
> If we increase `TASK_ATTEMPT_ID_MAX_LENGTH`, it will cause more GC.

why

**Comment by jerqi** [comment]
> > If we increase `TASK_ATTEMPT_ID_MAX_LENGTH`, it will cause more GC.
> 
> why

Because the implement of Roaringbitmap. If we increase the length of `TASK_ATTEMPT_ID_MAX_LENGTH`, it will occupy more memory.

**Comment by zuston** [comment]
Sorry, I can’t get your thought. Could you help give a link of roaringbitmap about this?

**Comment by jerqi** [comment]
You can see https://arxiv.org/pdf/1402.6407.pdf .

**Comment by zuston** [comment]
> You can see https://arxiv.org/pdf/1402.6407.pdf .

Thanks. Let me take a look later. 

And for this problem, how about decreasing partitionMaxLen to 20 by default, and the reduced length is given to seqNo?

**Comment by jerqi** [comment]
> > You can see https://arxiv.org/pdf/1402.6407.pdf .
> 
> Thanks. Let me take a look later.
> 
> And for this problem, how about decreasing partitionMaxLen to 20 by default, and the reduced length is given to seqNo?

Maybe, we should have more discussion about this problem.

**Comment by zuston** [comment]
BTW, if we could make this length as a client side config, this will more flexible. WDYT? @jerqi .

**Comment by jerqi** [comment]
Make sense.

**Comment by zuston** [comment]
@awdavidson 

In PR #749, @advancedxy propose a new possibility to solving this. refer https://github.com/apache/incubator-uniffle/pull/749#issuecomment-1486540366

But for my thought, it sounds good to me. But in the current codebase, the taskAttemptId must be a real value, some code will retrieve this from this blockId and do some validation. I'm not sure whether removing this part or optimize.

> What’s the feasibility of using a combination of the stageId and taskId? What else can be done?

I'm not sure what do you mean.  Could you help describe more?

**Comment by awdavidson** [comment]
> @awdavidson
> 
> In PR #749, @advancedxy propose a new possibility to solving this. refer [#749 (comment)](https://github.com/apache/incubator-uniffle/pull/749#issuecomment-1486540366)
> 
> But for my thought, it sounds good to me. But in the current codebase, the taskAttemptId must be a real value, some code will retrieve this from this blockId and do some validation. I'm not sure whether removing this part or optimize.
> 
> > What’s the feasibility of using a combination of the stageId and taskId? What else can be done?
> 
> I'm not sure what do you mean. Could you help describe more?

@zuston Thank you for the reply, yes I did notice that proposal and I agree with comments such as 

```
Also, I'm more concerned about the compatibility issues. To make the length of fields in block ids adaptive itself is appealing,
it's just that we really have to think it through and make sure it wouldn't break things.
```
In my opinion exposing this to the end user introduces risk and makes the assumption they know what they are doing. It would be ideal to have a implementation that just works without the end user needing to worry.

Sure, so firstly, I mean the taskId index. Currently taskId increments across the life of the application and represents the total number of tasks, which for uniffle has an upper limit.

```
Can't support taskAttemptId[3786263], the max value should be 2097151
```

taskId index is reset in each stage as the index relates to only the task set of the stage. Using a combination of the stageId and taskId index would help increase the upper limit especially for applications running iterative algorithms, rather than have `[1, ..., 20]` you would have `[1_1, .. , 1_10, 2_1, .. , 2_10]`. The blockId would be made up of something like: `long partitionId, long stageId, long taskIdIndex, long atomicInt`

**Comment by awdavidson** [comment]
> > > @awdavidson
> > > In PR #749, @advancedxy propose a new possibility to solving this. refer [#749 (comment)](https://github.com/apache/incubator-uniffle/pull/749#issuecomment-1486540366)
> > > But for my thought, it sounds good to me. But in the current codebase, the taskAttemptId must be a real value, some code will retrieve this from this blockId and do some validation. I'm not sure whether removing this part or optimize.
> > > > What’s the feasibility of using a combination of the stageId and taskId? What else can be done?
> > > 
> > > 
> > > I'm not sure what do you mean. Could you help describe more?
> > 
> > 
> > @zuston Thank you for the reply, yes I did notice that proposal and I agree with comments such as
> > ```
> > Also, I'm more concerned about the compatibility issues. To make the length of fields in block ids adaptive itself is appealing,
> > it's just that we really have to think it through and make sure it wouldn't break things.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > In my opinion exposing this to the end user introduces risk and makes the assumption they know what they are doing. It would be ideal to have a implementation that just works without the end user needing to worry.
> > Sure, so firstly, I mean the taskId index. Currently taskId increments across the life of the application and represents the total number of tasks, which for uniffle has an upper limit.
> > ```
> > Can't support taskAttemptId[3786263], the max value should be 2097151
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > taskId index is reset in each stage as the index relates to only the task set of the stage. Using a combination of the stageId and taskId index would help increase the upper limit especially for applications running iterative algorithms, rather than have `[1, ..., 20]` you would have `[1_1, .. , 1_10, 2_1, .. , 2_10]`. The blockId would be made up of something like: `long partitionId, long stageId, long taskIdIndex, long atomicInt`
> 
> task id is not identical in one stage. There could be duplicate tasks due to task retry or task speculation. We rely that info to deduplicate shuffle data in the shuffle read part.

A combination of task id and attempt number would be unique within a stage? Obviously you can have stage retries so I guess that would have to be taken into consideration.


Or uniffle should shuffle taskAttemptId to Long.MAX_VALUE?

**Comment by connorlwilkes** [comment]
@advancedxy  any more thoughts on this?

**Comment by EnricoMi** [comment]
Is this BlockID opaque to the RSS, or does the RSS extract and use the `partitionId`, `taskAttemptId` and `sequenceNumber` of a BlockID?

**Comment by jerqi** [comment]
> Is this BlockID opaque to the RSS, or does the RSS extract and use the `partitionId`, `taskAttemptId` and `sequenceNumber` of a BlockID?

RSS extracts and uses the partitionId, taskAttemptId and sequenceNumber of a BlockID

**Comment by EnricoMi** [comment]
Can you point me to the code, I could not find that.

**Comment by EnricoMi** [comment]
Given `partitionId` is 32 bit and `taskAttemptId` even 64 bit, there is always a chance that the BlockID overflows, no matter how we balance the different parts. Can you elaborate a bit on the design decision taken to have BlockID contain those parts, and in favour of what benefits you sacrifice large number of partitions or tasks.

I am happy to work out alternatives and contribute, but a bit of background knowledge is needed here. Maybe comments in on Github is not the best format to elaborate on this issue. Happy to chat elsewhere.

**Comment by jerqi** [comment]
> Given `partitionId` is 32 bit and `taskAttemptId` even 64 bit, there is always a chance that the BlockID overflows, no matter how we balance the different parts. Can you elaborate a bit on the design decision taken to have BlockID contain those parts, and in favour of what benefits you sacrifice large number of partitions or tasks.
> 
> I am happy to work out alternatives and contribute, but a bit of background knowledge is needed here. Maybe comments in on Github is not the best format to elaborate on this issue. Happy to chat elsewhere.

OK.  We have a slack channel or we can use a google meeting to communicate online.

**Comment by zuston** [comment]
Hey @EnricoMi Do you want to optimize this part? I prefer the length config could be held in the client firstly. After that, we could do more optimization about this.

Because the current server and client will access this conf at the same time and supposed that it will not changed. This is not proper.

**Comment by EnricoMi** [comment]
> I found some apps will fail due to the exceeding the  MAX_SEQUENCE_NO range

@zuston which Spark version were you using?

**Comment by zhengchenyu** [comment]
I also encountered the problem of block id overflow.
The block id is used to verify whether the accurate block set is obtained. I think we can get the sequence id from shuffle server. It will be almost impossible for overflow to occur.
But task id, task attempt id and partition id should be moved from block id.

**Comment by EnricoMi** [comment]
Here is a suggestion to make the bit lengths configurable: #1528

Here is a suggestion to effectively make those ids not overflow any more (for Spark): #1512

**Comment by EnricoMi** [comment]
Fix #1529 has been merged into master. That should mitigate the task attempt id overflow for Spark3.
